[MRG+2] DOC build pdf when master or version branch

waterponey opened the pull request:
#### Reference Issue

Takeover of PR #7671 for issue #7543 
#### What does this implement/fix? Explain your changes.

This modify the circleCI build to make it build pdf documentation when branch is master or a version number.

waterponey committed:
DOC build pdf when master or version branch

DOC build pdf when master or version branch

jnothman commented:
LGTM; at least worth trying!

waterponey commented:
I'm using two github account to check the behaviour after PR. Let me get back at you after that's checked.

lesteve commented on build_tools/circle/build_doc.sh:
```diff
@@ -51,5 +52,11 @@ source activate testenv
 # Build and install scikit-learn in dev mode
 python setup.py develop
 
+if [[ "$CIRCLE_BRANCH" =~ ^master$|^[0-9]+\.[0-9]+\.X$ ]]
```
Any reason why the -z "$CI_PULL_REQUEST" has disappeared from my latest comment [here](https://github.com/scikit-learn/scikit-learn/pull/7671#issuecomment-255779852).

> My understanding is that we should never build the pdf in pull requests because it takes too much time

My expectation is that CIRCLE_BRANCH is going to be master in a PR (because the PR is targeting master). I may be wrong about that.

lesteve replied:
Actually thinking about it, I remember now that CircleCI builds in branches and not (as you kind of would expect from Travis) on the result of the merge of the PR into master, so CIRCLE_BRANCH may be the name of the branch in the fork.

Still it's better to make sure that we never build the pdf in a PR, for example some beginners create PRs from their master branch sometimes.

amueller replied:
Hm but it would allow us to build pdf on PR branches if we want to.

waterponey replied:
Indeed, from what I understand, when you do a PR, the build is done on the branch once. In this first build, we'll have CIRCLE_BRANCH = "name_of_the_branch_on_your_fork" and CI_PULL_REQUEST="something"
Then I think, when you merge it to master, this generates another build which will be build from merged master with CI_PULL_REQUEST = nothing. That's what I'm trying to check right now but I keep having time out errors and error from Travis for some reasons I don't get...

lesteve replied:
> Hm but it would allow us to build pdf on PR branches if we want to.

I assume we wanted to mirror the behaviour of #7671, i.e. only build the pdf doc on pushes into master or into release branches and never inside a PR.

amueller replied:
yeah probably... depends on how long it takes.

amueller replied:
optipng certainly takes too long for doing it in branches.

amueller commented:
What happened to optipng? But I'm happy to merge this one first. LGTM

waterponey commented:
@amueller right now, I have a lot of timeout issues on my side when trying to build on circleCI so I'd rather do another PR later after I'm sure this PR is working well since I'm afraid optipng would take more time (that's what I understand from your previous discussion).

amueller commented:
ok fair.

lesteve commented:
Have you tried to run `make pdf` in your branch locally to make sure it succeeds ? Looking at your branch you don't have 387f25c6942e9cbd86370241f0e13886e8d94e07 (from 6 days ago), which fixed the latex build (confusingly and in contrary to TravisCI CircleCI does the build on your fork branch and not on the result of the merge of your branch into master).

What may happen is that you have a latex error and you end up hitting the usual latex interactive prompt in case of error asking you to type `x` to exit. That would be a reason for the time-out.

lesteve commented:
It looks like you don't have the raised (to 1 hour) timeout either. I would try to rebase your branch on master to see whether your problem just disappear ...

waterponey commented:
Ok, I'll try that ASAP. Thanks for the help !

waterponey commented:
Alright it seems to work as expected after a merge (it uses make dist instead of make html when it's a merge on master but not on PR). But it still times out so I have to take care of that...

waterponey committed:
Test for latex pdf generation [doc build]

lesteve commented:
> But it still times out so I have to take care of that...

Do you have a log with the timeout somewhere that we could look at? Also it would be great to run make dist locally to have some idea of how long it takes. If everything else fails, there is a way to set up SSH access in a CircleCI so we could try that.

waterponey commented:
I think I fixed it. I had two different errors. One were due to the mldata returning error 500. I didn't fix it because it works sometimes (and once it works, I think it caches the data so we're okay). But I'll open an issue on that since it is troublesome.

The other issue is that sphinx needed more latex dependencies so I added them in the build.sh.

I ran it on circleCI on my own account and it takes 1 hour and 10 minutes (compared with the 38 minutes for the ``make html`` only.

waterponey committed:
Fixup doc generation

jnothman commented:
Thanks! This is before including optipng, though? So the time is to vastly increase.

waterponey commented:
Yes. it's without optipng. I think we should think a bit more carefully before including it since it might be taking a bit too long with optipng on to build on each push to master. Maybe we could do something like use a [make pdf] tag that would trigger a specific pdf build ? We should discuss that on a separate issue or PR I think.

lesteve commented:
Since it takes 30 minutes more, I'd be in favour of making sure that we never build the pdf in PRs as I said in https://github.com/scikit-learn/scikit-learn/pull/7747#discussion_r84930115.

lesteve commented:
Other than that it's great that you got it working!

waterponey committed:
fixup adding is PR check

lesteve commented:
LGTM. Once this is merged we should monitor the build of the merge into master to make sure the pdf doc is built and there is no error.

waterponey commented:
I can monitor it on my own master first

waterponey commented:
It works (and took 40 minutes only)

jnothman commented:
I'm okay with merging this without optipng, but optipng does make a big
difference to PDF file size, and no doubt to overall bandwidth to the
website. There may also be ways to reduce duplicated work between HTML and
PDF builders, reducing runtime, but I suspect they're hacky.

We should always consider the solution that for the very occasional task of
updating PDF documentation for a release (i.e. not master) this should be
done by hand, or by hook to a private server.

On 28 October 2016 at 23:49, waterponey <notifications@github.com> wrote:

> It works (and took 40 minutes only)
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7747#issuecomment-256911893>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz60A7uofsYNUveYw7iR5rGJYbL8cGks5q4e9DgaJpZM4Kfnm2>
> .
>

waterponey commented:
Does a lot of people download the pdf documentation ?

jnothman commented:
No, but if we don't have it, people request it. :)

On 29 October 2016 at 23:06, waterponey <notifications@github.com> wrote:

> Does a lot of people download the pdf documentation ?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/7747#issuecomment-257088008>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6wGMefMt4vQem1W2W9u1uW8WYBzYks5q4zapgaJpZM4Kfnm2>
> .
>

jnothman commented:
Which means, I suppose, that it's better there and big than absent!

On 29 October 2016 at 23:22, Joel Nothman <joel.nothman@gmail.com> wrote:

> No, but if we don't have it, people request it. :)
>
> On 29 October 2016 at 23:06, waterponey <notifications@github.com> wrote:
>
>> Does a lot of people download the pdf documentation ?
>>
>> —
>> You are receiving this because you commented.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/scikit-learn/scikit-learn/pull/7747#issuecomment-257088008>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AAEz6wGMefMt4vQem1W2W9u1uW8WYBzYks5q4zapgaJpZM4Kfnm2>
>> .
>>
>
>

jnothman commented:
Thanks! Let's see how this goes...
